sharktuner/boo_tuner.py

```diff
@@ -0,0 +1,124 @@
+import shlex
+from pathlib import Path
+import os
+import tempfile
+import subprocess
+import shutil
+import traceback
+
+from iree.turbine.kernel.boo import runtime as boo_runtime
+from iree.turbine.kernel.boo.driver.launch import get_launchable
+from iree.turbine.kernel.boo.op_exports.conv import ConvParser as mio
+
+from model_tuner.model_tuner import main as tuner_main
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument("--commands-file", type=str, help="read commands from file")
+    parser.add_argument("--output-td-spec", type=Path, default="tuning-spec.mlir")
+    parser.add_argument("--starter-td-spec", type=Path)
+    parser.add_argument("--num-candidates", type=int, default=100)
+    parser.add_argument("--devices", type=str, default="hip://0")
+    parser.add_argument(
+        "--tmp-dir", type=str, default="", help="directory to save temporary files"
+    )
+    args, extra_cli_args = parser.parse_known_args()
+
+    if args.commands_file:
+        with open(args.commands_file) as f:
+            mio_file_args = [
+                shlex.split(s) for s in f.readlines() if not s.startswith("#")
+            ]
+    else:
+        mio_file_args = [[]]  # use CLI arguments
+
+    starter_td_spec: Path | None = args.starter_td_spec
+    for idx, file_args in enumerate(mio_file_args):
+        cli_args = file_args + extra_cli_args
+        print(f">>> ({idx+1}/{len(mio_file_args)}) {shlex.join(cli_args)}")
+        parser = mio.get_miopen_parser()
+        parser.add_argument("-t")
+        parser.add_argument("--iter")
+        sig = mio.get_signature(parser.parse_args(cli_args))
+        if args.tmp_dir:
+            tmp_dir = Path(args.tmp_dir)
+            # Make sure directory is empty.
+            shutil.rmtree(tmp_dir, ignore_errors=True)
+            os.mkdir(tmp_dir)
+        else:
+            tmp_dir = Path(tempfile.mkdtemp(dir=".", prefix="boo-tuner-"))
+        boo_cache_dir = tmp_dir / "boo_cache"
+
+        # Run BOO compilation an extract source IR.
+        with boo_runtime.use_cache_dir(boo_cache_dir):
+            get_launchable(sig)(*sig.get_sample_args(device="cuda", seed=123))
+        [op_cache_dir] = os.listdir(boo_cache_dir)
+        [compile_command_path] = [
+            f
+            for f in os.listdir(boo_cache_dir / op_cache_dir)
+            if f.startswith("compile_command")
+        ]
+        with open(boo_cache_dir / op_cache_dir / compile_command_path) as f:
+            compile_command = f.read().strip()
+
+        # Re-compile to dump dispatches.
+        bench_dir = tmp_dir / "bench"
+        compile_args = shlex.split(compile_command) + [
+            "--iree-codegen-tuning-spec-path=",  # disable tuning spec flag as it prevents 'iree-config-add-tuner-attributes' from working.
+            "--iree-config-add-tuner-attributes",
+            "--iree-hal-dump-executable-benchmarks-to",
+            str(bench_dir),
+            "-o",
+            "/dev/null",
```

Comment:
Does this work on windows? We don't currently have any CI runners to check, but eventually we would like to support it too.